### ISSUE TITLE
utf8_utils.hpp -> codepoint_to_utf8  bug 

### ISSUE THREAD

**Issue by bizehao** [cycfi/infra#16]
` inline std::string codepoint_to_utf8(char32_t codepoint)`
 `  {`
          &emsp; &emsp;`std::string result{ 8 };`
          &emsp; &emsp;`detail::codepoint_to_utf8(codepoint, &result[0]);`
          &emsp; &emsp; `return result;`
  ` }` 

In win10 vs2022 is 4 bytes, Therefore, when inputting Chinese characters, it will lead to subsequent processing errors.

Here are my modifications:

` inline std::string codepoint_to_utf8(char32_t codepoint)`
 `  {`
      &emsp; &emsp; `char cache[8];`
      &emsp; &emsp; ` auto result = detail::codepoint_to_utf8(codepoint, cache);`
      &emsp; &emsp;`return result;`
  ` }`

**Comment by djowel** [comment]
> In win10 vs2022 is 4 bytes, Therefore, when inputting Chinese characters, it will lead to subsequent processing errors.

I don't understand what you mean by "In win10 vs2022 is 4 bytes". It will be good to provide an example. Please file a pull request with detailed information and explanation.

**Comment by bizehao** [comment]
He happened in [https://github.com/cycfi/elements/blob/artist_port/examples/text_edit/main.cpp](https://github.com/cycfi/elements/blob/artist_port/examples/text_edit/main.cpp). <br> When Input Chinese with keyboard in `basic_text_box`, Exception thrown: read access violation. <br>When I modify the above code, the input is normal.

**Comment by djowel** [comment]
Can you give me the Chinese characters you typed so I can try?

Also, please explain what you mean by "In win10 vs2022 is 4 bytes". I don't understand how that affects the code.

**Comment by bizehao** [comment]
Chinese characters：的
<br/>
Because directly reported the wrong memory, I changed char [8],it is good,so I guess that is char [4]

**Comment by djowel** [comment]
I have no problem with it:

![image](https://user-images.githubusercontent.com/568094/150470828-851ac3c1-f924-4c12-a139-4e3145bb9f52.png)

**Comment by djowel** [comment]
> Because directly reported the wrong memory, I changed char [8],it is good,so I guess that is char [4]

Where did you find `char [4]`? Still does not make sense.

**Comment by bizehao** [comment]
Is your text input to the keyboard?

**Comment by djowel** [comment]
Which branch are you using? master? develop?

**Comment by djowel** [comment]
> Is your text input to the keyboard?

It does not matter. Have you tried copy/paste?

**Comment by bizehao** [comment]
artist _ port branch. I didn't tried the copy paste, waiting for me to go back at night and try it.

**Comment by djowel** [comment]
I see. Can you test `develop` branch?

Next time, please provide as much information as possible, so we don't waste time.

**Comment by bizehao** [comment]
ok,I will test it together.

**Comment by djowel** [comment]
> ok,I will test it together.

Thank you

**Comment by bizehao** [comment]
Can I ask you? Why do you use std :: string {8}?

**Comment by djowel** [comment]
> Can I ask you? Why do you use std :: string {8}?

Why not?

**Comment by djowel** [comment]
Anyway, I tried it again with `artist_port`. I am not having any problems. 

If you run it again, please give me a stack trace, so I can know which functions is calling `codepoint_to_utf8`. I don't see any problem with the code with `std :: string {8}`. Can you explain why you think it is a bug?

**Comment by bizehao** [comment]
Change to std::string {9} std::string {10} is ok?

**Comment by bizehao** [comment]
I didn't think he was a mistake. It is the error that I am running, I may be related to my system, my Chinese version of Win10

**Comment by djowel** [comment]
> Change to std::string {9} std::string {10} is ok?

Any size will do. It can be `std::string{16}`. 

But what I do not understand is why `char cache[8];` is OK. It has the same size: 8.

**Comment by bizehao** [comment]
sorry，Forgot one thing, it is the CPP20 version I used, I will go home overall test over again.

**Comment by djowel** [comment]
Can you try `std::string(8, ' ')` ?

**Comment by djowel** [comment]
> sorry，Forgot one thing, it is the CPP20 version I used, I will go home overall test over again.

That might be it! I am using C++17. Please try `std::string(8, ' ')`

**Comment by bizehao** [comment]
ok,Perhaps it is,At work,I will conclusion to you at night.

**Comment by djowel** [comment]
Anyway, I think I'll accept your original solution. I think I am beginning to understand the issue. Please file a PR so we can conclude this issue.

```
inline std::string codepoint_to_utf8(char32_t codepoint)
{
   char cache[8];
   return detail::codepoint_to_utf8(codepoint, cache);
}
```

**Comment by bizehao** [comment]
The problem I found is the one you said because of the problem of {}. I went through the initialization list
Use {}, get into initializer_list. When I change it to (), it will be normal.

**Comment by bizehao** [comment]
ok, I'll submit a pr.

**Comment by redtide** [comment]
was this fixed by #17?
